File: src/Python/src/vladimir_reshetnikov/data_structures/finger_tree/rrb_vector.py
"""Persistent relaxed radix-balanced vectors with 32-way branching."""

from __future__ import annotations

import sys
from collections.abc import Iterable, Iterator, Sequence
from dataclasses import dataclass
from typing import Generic, TypeVar, overload

T = TypeVar("T")

_BRANCH_FACTOR = 32
_RADIX_BITS = 5


@dataclass(frozen=True, slots=True)
class _Leaf(Generic[T]):
    items: tuple[T, ...]
    count: int
    height: int = 0

    def __init__(self, items: Iterable[T]) -> None:
        materialized = tuple(items)
        if not 1 <= len(materialized) <= _BRANCH_FACTOR:
            raise ValueError("Invalid RRB leaf size.")
        object.__setattr__(self, "items", materialized)
        object.__setattr__(self, "count", len(materialized))
        object.__setattr__(self, "height", 0)


@dataclass(frozen=True, slots=True)
class _Branch(Generic[T]):
    children: tuple[_Node[T], ...]
    height: int
    count: int
    cumulative_sizes: tuple[int, ...] | None

    def __init__(self, children: Iterable[_Node[T]]) -> None:
        materialized = tuple(children)
        if not 1 <= len(materialized) <= _BRANCH_FACTOR:
            raise ValueError("Invalid RRB branch factor.")
        child_height = materialized[0].height
        if any(child.height != child_height for child in materialized):
            raise ValueError("RRB siblings must have equal heights.")
        height = child_height + 1
        count = 0
        for child in materialized:
            count = _checked_count(count, child.count)
        object.__setattr__(self, "children", materialized)
        object.__setattr__(self, "height", height)
        object.__setattr__(self, "count", count)
        object.__setattr__(
            self,
            "cumulative_sizes",
            None if _has_regular_layout(materialized, height) else _build_sizes(materialized),
        )

    def find_child(self, index: int) -> tuple[int, int]:
        if index < 0 or index >= self.count:
            raise IndexError("RRB child index is outside the branch.")
        if self.cumulative_sizes is None:
            capacity = 1 << (self.height * _RADIX_BITS)
            child_index = index // capacity
            return child_index, child_index * capacity
        low = 0
        high = len(self.cumulative_sizes) - 1
        while low < high:
            middle = (low + high) // 2
            if index < self.cumulative_sizes[middle]:
                high = middle
            else:
                low = middle + 1
        return low, 0 if low == 0 else self.cumulative_sizes[low - 1]


_Node = _Leaf[T] | _Branch[T]


def _checked_count(left: int, right: int) -> int:
    result = left + right
    if result > sys.maxsize:
        raise OverflowError("An RRB vector cannot exceed sys.maxsize elements.")
    return result


def _has_regular_layout(children: Sequence[_Node[T]], height: int) -> bool:
    capacity = 1 << (height * _RADIX_BITS)
    return (
        all(child.count == capacity for child in children[:-1]) and children[-1].count <= capacity
    )


def _build_sizes(children: Sequence[_Node[T]]) -> tuple[int, ...]:
    sizes: list[int] = []
    count = 0
    for child in children:
        count = _checked_count(count, child.count)
        sizes.append(count)
    return tuple(sizes)


def _get_node(root: _Node[T], initial_index: int) -> T:
    node = root
    index = initial_index
    while isinstance(node, _Branch):
        child_index, before = node.find_child(index)
        index -= before
        node = node.children[child_index]
    return node.items[index]


def _set_node(node: _Node[T], index: int, value: T) -> _Node[T]:
    if isinstance(node, _Leaf):
        if node.items[index] is value:
            return node
        items = list(node.items)
        items[index] = value
        return _Leaf(items)
    child_index, before = node.find_child(index)
    current = node.children[child_index]
    child = _set_node(current, index - before, value)
    if child is current:
        return node
    children = list(node.children)
    children[child_index] = child
    return _Branch(children)


def _partition(children: Sequence[_Node[T]]) -> tuple[_Node[T], ...]:
    if len(children) <= _BRANCH_FACTOR:
        return (_Branch(children),)
    split = len(children) // 2
    return _Branch(children[:split]), _Branch(children[split:])


def _concat_same_height(left: _Node[T], right: _Node[T]) -> tuple[_Node[T], ...]:
    if isinstance(left, _Leaf):
        if not isinstance(right, _Leaf):
            raise AssertionError("RRB height mismatch.")
        if left.count == _BRANCH_FACTOR and right.count == _BRANCH_FACTOR:
            return left, right
        combined = left.items + right.items
        if len(combined) <= _BRANCH_FACTOR:
            return (_Leaf(combined),)
        split = len(combined) // 2
        return _Leaf(combined[:split]), _Leaf(combined[split:])
    if not isinstance(right, _Branch):
        raise AssertionError("RRB height mismatch.")
    boundary = _concat_same_height(left.children[-1], right.children[0])
    return _partition((*left.children[:-1], *boundary, *right.children[1:]))


def _concat_nodes(left: _Node[T], right: _Node[T]) -> tuple[_Node[T], ...]:
    if left.height == right.height:
        return _concat_same_height(left, right)
    if left.height > right.height:
        if not isinstance(left, _Branch):
            raise AssertionError("RRB height mismatch.")
        boundary = _concat_nodes(left.children[-1], right)
        return _partition((*left.children[:-1], *boundary))
    if not isinstance(right, _Branch):
        raise AssertionError("RRB height mismatch.")
    boundary = _concat_nodes(left, right.children[0])
    return _partition((*boundary, *right.children[1:]))


def _build_same_height(nodes: Sequence[_Node[T]]) -> _Node[T] | None:
    return None if not nodes else _Branch(nodes)


def _split_node(node: _Node[T], index: int) -> tuple[_Node[T] | None, _Node[T] | None]:
    if index == 0:
        return None, node
    if index == node.count:
        return node, None
    if isinstance(node, _Leaf):
        return _Leaf(node.items[:index]), _Leaf(node.items[index:])
    child_index, before = node.find_child(index)
    child_left, child_right = _split_node(node.children[child_index], index - before)
    left = (*node.children[:child_index], *((child_left,) if child_left is not None else ()))
    right = (
        *((child_right,) if child_right is not None else ()),
        *node.children[child_index + 1 :],
    )
    return _build_same_height(left), _build_same_height(right)


def _build_level(nodes: Sequence[_Node[T]]) -> _Node[T]:
    if not nodes:
        raise ValueError("Cannot build an empty RRB level.")
    level = list(nodes)
    while len(level) > 1:
        level = [
            _Branch(level[index : index + _BRANCH_FACTOR])
            for index in range(0, len(level), _BRANCH_FACTOR)
        ]
    return level[0]


def _normalize_root(candidate: _Node[T] | None) -> _Node[T] | None:
    node = candidate
    while isinstance(node, _Branch) and len(node.children) == 1:
        node = node.children[0]
    return node


def _iterate(root: _Node[T]) -> Iterator[T]:
    stack = [root]
    while stack:
        node = stack.pop()
        if isinstance(node, _Leaf):
            yield from node.items
        else:
            stack.extend(reversed(node.children))


@dataclass(frozen=True, slots=True)
class RrbPop(Generic[T]):
    value: T
    rest: RrbVector[T]


@dataclass(frozen=True, slots=True)
class RrbVectorSplit(Generic[T]):
    left: RrbVector[T]
    right: RrbVector[T]


@dataclass(frozen=True, slots=True)
class RrbVectorStatistics:
    count: int
    height: int
    leaf_count: int
    branch_count: int
    regular_branch_count: int
    relaxed_branch_count: int
    minimum_leaf_length: int
    maximum_leaf_length: int
    minimum_branching_factor: int
    maximum_branching_factor: int


class RrbVector(Generic[T]):
    """Immutable relaxed radix-balanced vector with 32-way branches."""

    __slots__ = ("_root",)

    def __init__(self, root: _Node[T] | None = None) -> None:
        self._root = _normalize_root(root)

    @classmethod
    def empty(cls) -> RrbVector[T]:
        return cls()

    @classmethod
    def from_iterable(cls, values: Iterable[T]) -> RrbVector[T]:
        if isinstance(values, RrbVector):
            return values
        return RrbVectorBuilder[T]().append_all(values).to_immutable()

    @classmethod
    def builder(cls) -> RrbVectorBuilder[T]:
        return RrbVectorBuilder()

    def __len__(self) -> int:
        return 0 if self._root is None else self._root.count

    @property
    def size(self) -> int:
        return len(self)

    @property
    def is_empty(self) -> bool:
        return self._root is None

    @property
    def height(self) -> int:
        return 0 if self._root is None else self._root.height

    def get(self, index: int) -> T | None:
        if index < 0 or index >= len(self):
            return None
        if self._root is None:
            raise AssertionError("A validated nonempty vector has no root.")
        return _get_node(self._root, index)

    @overload
    def __getitem__(self, index: int) -> T: ...

    @overload
    def __getitem__(self, index: slice) -> RrbVector[T]: ...

    def __getitem__(self, index: int | slice) -> T | RrbVector[T]:
        if isinstance(index, slice):
            start, stop, step = index.indices(len(self))
            if step != 1:
                return RrbVector.from_iterable(self.to_list()[index])
            split_start = self.split_at(start)
            if split_start is None:
                raise AssertionError("Normalized slice start failed.")
            split_length = split_start.right.split_at(stop - start)
            if split_length is None:
                raise AssertionError("Normalized slice length failed.")
            return split_length.left
        normalized = index + len(self) if index < 0 else index
        value = self.get(normalized)
        if normalized < 0 or normalized >= len(self):
            raise IndexError("RRB vector index out of range.")
        return value  # type: ignore[return-value]

    def front(self) -> T | None:
        return self.get(0)

    def back(self) -> T | None:
        return self.get(len(self) - 1)

    def append(self, value: T) -> RrbVector[T]:
        return self.concat(RrbVector(_Leaf((value,))))

    def prepend(self, value: T) -> RrbVector[T]:
        return RrbVector(_Leaf((value,))).concat(self)

    def set_item(self, index: int, value: T) -> RrbVector[T] | None:
        if index < 0 or index >= len(self):
            return None
        if self._root is None:
            raise AssertionError("A validated nonempty vector has no root.")
        root = _set_node(self._root, index, value)
        return self if root is self._root else RrbVector(root)

    def concat(self, other: RrbVector[T]) -> RrbVector[T]:
        if self._root is None:
            return other
        if other._root is None:
            return self
        _checked_count(len(self), len(other))
        roots = _concat_nodes(self._root, other._root)
        return RrbVector(roots[0] if len(roots) == 1 else _Branch(roots))

    def split_at(self, index: int) -> RrbVectorSplit[T] | None:
        if index < 0 or index > len(self):
            return None
        if index == 0:
            return RrbVectorSplit(RrbVector.empty(), self)
        if index == len(self):
            return RrbVectorSplit(self, RrbVector.empty())
        if self._root is None:
            raise AssertionError("A validated nonempty vector has no root.")
        left, right = _split_node(self._root, index)
        return RrbVectorSplit(RrbVector(left), RrbVector(right))

    def insert_at(self, index: int, value: T) -> RrbVector[T] | None:
        return self.insert_range(index, (value,))

    def insert_range(self, index: int, values: Iterable[T]) -> RrbVector[T] | None:
        split = self.split_at(index)
        if split is None:
            return None
        middle = RrbVector.from_iterable(values)
        return self if middle.is_empty else split.left.concat(middle).concat(split.right)

    def remove_at(self, index: int) -> RrbVector[T] | None:
        return self.remove_range(index, 1)

    def remove_range(self, index: int, count: int) -> RrbVector[T] | None:
        if index < 0 or count < 0 or index + count > len(self):
            return None
        if count == 0:
            return self
        first = self.split_at(index)
        if first is None:
            raise AssertionError("Validated first split failed.")
        second = first.right.split_at(count)
        if second is None:
            raise AssertionError("Validated second split failed.")
        return first.left.concat(second.right)

    def try_remove_last(self) -> RrbPop[T] | None:
        if self.is_empty:
            return None
        value = self.back()
        rest = self.remove_at(len(self) - 1)
        if rest is None:
            raise AssertionError("Validated last removal failed.")
        return RrbPop(value, rest)  # type: ignore[arg-type]

    def clear(self) -> RrbVector[T]:
        return self if self.is_empty else RrbVector.empty()

    def to_builder(self) -> RrbVectorBuilder[T]:
        return RrbVectorBuilder(self)

    def to_list(self) -> list[T]:
        return list(self)

    def shares_root_with(self, other: RrbVector[T]) -> bool:
        return self._root is other._root

    def shared_leaf_count(self, other: RrbVector[T]) -> int:
        leaves: set[int] = set()
        stack = [] if self._root is None else [self._root]
        while stack:
            node = stack.pop()
            if isinstance(node, _Leaf):
                leaves.add(id(node))
            else:
                stack.extend(node.children)
        count = 0
        probe = [] if other._root is None else [other._root]
        while probe:
            node = probe.pop()
            if isinstance(node, _Leaf):
                count += id(node) in leaves
            else:
                probe.extend(node.children)
        return count

    def validate_structure(self) -> RrbVectorStatistics:
        if self._root is None:
            return RrbVectorStatistics(0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
        leaf_count = branch_count = regular = relaxed = 0
        min_leaf = _BRANCH_FACTOR
        max_leaf = 0
        min_branch = _BRANCH_FACTOR
        max_branch = 0

        def visit(node: _Node[T], is_root: bool) -> tuple[int, int]:
            nonlocal leaf_count, branch_count, regular, relaxed
            nonlocal min_leaf, max_leaf, min_branch, max_branch
            if isinstance(node, _Leaf):
                if not 1 <= node.count <= _BRANCH_FACTOR or node.count != len(node.items):
                    raise ValueError("Invalid RRB leaf size.")
                leaf_count += 1
                min_leaf = min(min_leaf, node.count)
                max_leaf = max(max_leaf, node.count)
                return node.count, 0
            if not 1 <= len(node.children) <= _BRANCH_FACTOR:
                raise ValueError("Invalid RRB branch factor.")
            if is_root and len(node.children) == 1:
                raise ValueError("An RRB root branch must have at least two children.")
            branch_count += 1
            min_branch = min(min_branch, len(node.children))
            max_branch = max(max_branch, len(node.children))
            if node.cumulative_sizes is None:
                regular += 1
            else:
                relaxed += 1
            results = [visit(child, False) for child in node.children]
            if any(height != results[0][1] for _, height in results):
                raise ValueError("RRB sibling height mismatch.")
            count = sum(child_count for child_count, _ in results)
            height = results[0][1] + 1
            if count != node.count or height != node.height:
                raise ValueError("Invalid RRB metadata.")
            expected_regular = _has_regular_layout(node.children, height)
            if (node.cumulative_sizes is None) != expected_regular:
                raise ValueError("Invalid RRB regularity metadata.")
            if node.cumulative_sizes is not None and node.cumulative_sizes != _build_sizes(
                node.children
            ):
                raise ValueError("Invalid RRB size table.")
            return count, height

        count, height = visit(self._root, True)
        return RrbVectorStatistics(
            count,
            height,
            leaf_count,
            branch_count,
            regular,
            relaxed,
            min_leaf,
            max_leaf,
            0 if branch_count == 0 else min_branch,
            max_branch,
        )

    def __iter__(self) -> Iterator[T]:
        return iter(()) if self._root is None else _iterate(self._root)


class RrbVectorBuilder(Generic[T]):
    """Mutable append staging surface that publishes immutable RRB snapshots."""

    __slots__ = ("_leaves", "_prefix", "_staged_count", "_tail", "_version")

    def __init__(self, prefix: RrbVector[T] | None = None) -> None:
        self._prefix = RrbVector.empty() if prefix is None else prefix
        self._leaves: list[_Node[T]] = []
        self._tail: list[T] = []
        self._staged_count = 0
        self._version = 0

    def __len__(self) -> int:
        return _checked_count(len(self._prefix), self._staged_count)

    @property
    def size(self) -> int:
        return len(self)

    def append(self, value: T) -> RrbVectorBuilder[T]:
        self._tail.append(value)
        self._staged_count += 1
        if len(self._tail) == _BRANCH_FACTOR:
            self._leaves.append(_Leaf(self._tail))
            self._tail = []
        self._version += 1
        return self

    def append_all(self, values: Iterable[T]) -> RrbVectorBuilder[T]:
        for value in values:
            self.append(value)
        return self

    def clear(self) -> RrbVectorBuilder[T]:
        if len(self) != 0:
            self._prefix = RrbVector.empty()
            self._leaves = []
            self._tail = []
            self._staged_count = 0
            self._version += 1
        return self

    def to_immutable(self) -> RrbVector[T]:
        if self._staged_count == 0:
            return self._prefix
        nodes = self._leaves if not self._tail else [*self._leaves, _Leaf(self._tail)]
        staged = RrbVector(_build_level(nodes))
        self._prefix = staged if self._prefix.is_empty else self._prefix.concat(staged)
        self._leaves = []
        self._tail = []
        self._staged_count = 0
        return self._prefix

    def __iter__(self) -> Iterator[T]:
        version = self._version
        for value in self.to_immutable():
            if version != self._version:
                raise RuntimeError("RRB builder modified during iteration.")
            yield value


__all__ = [
    "RrbPop",
    "RrbVector",
    "RrbVectorBuilder",
    "RrbVectorSplit",
    "RrbVectorStatistics",
]

File: src/OCaml/lib/finger_tree/rrb_vector.ml
(** Implementation of the persistent relaxed radix-balanced vector with 32-way branching.

    A genuine RRB tree, not a facade: leaves pack up to 32 elements in flat arrays, branches pack
    up to 32 children, and a branch whose children fill the 5-bit radix layout exactly is
    {e regular} and stores no size table — addressing it is pure shift-and-mask arithmetic. Only an
    irregular branch pays for cumulative sizes and answers by binary search, which is what makes
    O(log n) worst-case splits and concatenations affordable without a global occupancy rule.
    Concatenation rebalances only the seam between its operands, endpoint pushes are singleton
    concatenations, and everything is eager: indexing and point writes are O(log32 n) worst case,
    splits and concatenations O(log n) worst case, with no deferred work anywhere. *)

let branch_factor = 32
let radix_bits = 5

type 'element node =
  | Leaf of 'element array
  | Branch of {
      children : 'element node array;
      height : int;
      count : int;
      (* Cumulative child sizes, present exactly when the layout is irregular. A regular branch
         addresses arithmetically; storing a table there would be wasted space, and omitting it on
         an irregular branch would make radix arithmetic wrong. *)
      sizes : int array option;
    }

type 'element t = Empty | Root of 'element node
type 'element cursor = { cursor_snapshot : 'element t; cursor_position : int }
type statistics = { count : int; estimated_leaves : int; depth : int }

let node_count = function Leaf items -> Array.length items | Branch branch -> branch.count
let node_height = function Leaf _ -> 0 | Branch branch -> branch.height

(* The capacity of one child of a branch at the given height: 32^height, a power of two. *)
let child_capacity height = 1 lsl (height * radix_bits)

let has_regular_layout children height =
  let capacity = child_capacity height in
  let last = Array.length children - 1 in
  let regular = ref (node_count children.(last) <= capacity) in
  for index = 0 to last - 1 do
    if node_count children.(index) <> capacity then regular := false
  done;
  !regular

let build_sizes children =
  let sizes = Array.make (Array.length children) 0 in
  let total = ref 0 in
  Array.iteri
    (fun index child ->
      total := !total + node_count child;
      sizes.(index) <- !total)
    children;
  sizes

(* Builds a branch, deriving its height and count and deciding whether it needs a size table. *)
let make_branch children =
  let width = Array.length children in
  assert (width >= 1 && width <= branch_factor);
  let height = node_height children.(0) + 1 in
  assert (Array.for_all (fun child -> node_height child = height - 1) children);
  let count = Array.fold_left (fun total child -> total + node_count child) 0 children in
  let sizes = if has_regular_layout children height then None else Some (build_sizes children) in
  Branch { children; height; count; sizes }

(* The child covering the index, and how many elements precede that child. A regular branch
   computes both by radix arithmetic; an irregular one binary-searches its size table. *)
let find_child children height sizes index =
  match sizes with
  | None ->
      let shift = height * radix_bits in
      let child_index = index lsr shift in
      (child_index, child_index lsl shift)
  | Some sizes ->
      let low = ref 0 in
      let high = ref (Array.length children - 1) in
      while !low < !high do
        let middle = (!low + !high) / 2 in
        if index < sizes.(middle) then high := middle else low := middle + 1
      done;
      (!low, if !low = 0 then 0 else sizes.(!low - 1))

let rec get_node node index =
  match node with
  | Leaf items -> items.(index)
  | Branch branch ->
      let child_index, before = find_child branch.children branch.height branch.sizes index in
      get_node branch.children.(child_index) (index - before)

let rec set_node node index value =
  match node with
  | Leaf items ->
      let items = Array.copy items in
      items.(index) <- value;
      Leaf items
  | Branch branch ->
      let child_index, before = find_child branch.children branch.height branch.sizes index in
      let children = Array.copy branch.children in
      children.(child_index) <- set_node children.(child_index) (index - before) value;
      make_branch children

(* Wraps same-height siblings into one branch, or two when they overflow the branch factor. *)
let partition children =
  if Array.length children <= branch_factor then [| make_branch children |]
  else
    let split = Array.length children / 2 in
    [|
      make_branch (Array.sub children 0 split);
      make_branch (Array.sub children split (Array.length children - split));
    |]

let splice_children before boundary after =
  Array.concat [ before; boundary; after ]

(* Concatenates two same-height nodes, rebalancing only the seam between them and returning one or
   two nodes for the caller to wrap. Two full leaves are kept as they are; anything smaller merges
   or redistributes so the seam never accumulates fragmentation. *)
let rec concat_same_height left right =
  match (left, right) with
  | Leaf left_items, Leaf right_items ->
      let left_length = Array.length left_items in
      let right_length = Array.length right_items in
      if left_length = branch_factor && right_length = branch_factor then [| left; right |]
      else
        let combined = Array.append left_items right_items in
        if Array.length combined <= branch_factor then [| Leaf combined |]
        else
          let split = Array.length combined / 2 in
          [|
            Leaf (Array.sub combined 0 split);
            Leaf (Array.sub combined split (Array.length combined - split));
          |]
  | Branch left_branch, Branch right_branch ->
      let left_children = left_branch.children in
      let right_children = right_branch.children in
      let boundary =
        concat_same_height
          left_children.(Array.length left_children - 1)
          right_children.(0)
      in
      partition
        (splice_children
           (Array.sub left_children 0 (Array.length left_children - 1))
           boundary
           (Array.sub right_children 1 (Array.length right_children - 1)))
  | Leaf _, Branch _ | Branch _, Leaf _ -> assert false

(* Concatenates two nodes of any heights by descending the taller operand's seam edge. *)
let rec concat_nodes left right =
  if node_height left = node_height right then concat_same_height left right
  else if node_height left > node_height right then
    match left with
    | Branch branch ->
        let children = branch.children in
        let boundary = concat_nodes children.(Array.length children - 1) right in
        partition (splice_children (Array.sub children 0 (Array.length children - 1)) boundary [||])
    | Leaf _ -> assert false
  else
    match right with
    | Branch branch ->
        let children = branch.children in
        let boundary = concat_nodes left children.(0) in
        partition (splice_children [||] boundary (Array.sub children 1 (Array.length children - 1)))
    | Leaf _ -> assert false

(* Collapses single-child chains at the root; interior single-child branches left by splitting are
   legal and get merged away by later seam concatenations. *)
let rec normalize_root node =
  match node with
  | Branch branch when Array.length branch.children = 1 -> normalize_root branch.children.(0)
  | Leaf _ | Branch _ -> node

let wrap = function None -> Empty | Some node -> Root (normalize_root node)

let build_same_height children = if Array.length children = 0 then None else Some (make_branch children)

let rec split_node node index =
  if index = 0 then (None, Some node)
  else if index = node_count node then (Some node, None)
  else
    match node with
    | Leaf items ->
        ( Some (Leaf (Array.sub items 0 index)),
          Some (Leaf (Array.sub items index (Array.length items - index))) )
    | Branch branch ->
        let child_index, before = find_child branch.children branch.height branch.sizes index in
        let child_left, child_right = split_node branch.children.(child_index) (index - before) in
        let left =
          splice_children
            (Array.sub branch.children 0 child_index)
            (match child_left with None -> [||] | Some child -> [| child |])
            [||]
        in
        let right =
          splice_children [||]
            (match child_right with None -> [||] | Some child -> [| child |])
            (Array.sub branch.children (child_index + 1)
               (Array.length branch.children - child_index - 1))
        in
        (build_same_height left, build_same_height right)

(* --- public operations ----------------------------------------------------------------------- *)

let empty = Empty
let singleton value = Root (Leaf [| value |])
let length = function Empty -> 0 | Root node -> node_count node
let is_empty = function Empty -> true | Root _ -> false

(* Groups same-height nodes into branches level by level: an eager O(n) bulk build whose branches
   are all regular, so a freshly built vector addresses arithmetically everywhere. *)
let build_level nodes =
  let rec grow nodes =
    let width = Array.length nodes in
    if width = 1 then nodes.(0)
    else
      let groups = (width + branch_factor - 1) / branch_factor in
      grow
        (Array.init groups (fun group ->
             let start = group * branch_factor in
             make_branch (Array.sub nodes start (min branch_factor (width - start)))))
  in
  grow nodes

let of_list values =
  match values with
  | [] -> Empty
  | _ ->
      let items = Array.of_list values in
      let total = Array.length items in
      let leaves = (total + branch_factor - 1) / branch_factor in
      Root
        (build_level
           (Array.init leaves (fun leaf ->
                let start = leaf * branch_factor in
                Leaf (Array.sub items start (min branch_factor (total - start))))))

let nth index vector =
  match vector with
  | Empty -> None
  | Root node -> if index < 0 || index >= node_count node then None else Some (get_node node index)

let set index value vector =
  match vector with
  | Root node when index >= 0 && index < node_count node -> Ok (Root (set_node node index value))
  | Empty | Root _ -> Error "sequence update index is out of range"

let concat left right =
  match (left, right) with
  | Empty, other | other, Empty -> other
  | Root left_node, Root right_node -> (
      match concat_nodes left_node right_node with
      | [| single |] -> Root (normalize_root single)
      | pair -> Root (make_branch pair))

let append value vector = concat vector (singleton value)
let prepend value vector = concat (singleton value) vector

let split_at index vector =
  match vector with
  | Empty -> (Empty, Empty)
  | Root node ->
      if index <= 0 then (Empty, vector)
      else if index >= node_count node then (vector, Empty)
      else
        let left, right = split_node node index in
        (wrap left, wrap right)

let take count vector = fst (split_at count vector)
let drop count vector = snd (split_at count vector)

let insert index value vector =
  if index < 0 || index > length vector then Error "sequence insertion index is out of range"
  else
    let left, right = split_at index vector in
    Ok (concat (concat left (singleton value)) right)

let remove index vector =
  if index < 0 || index >= length vector then Error "sequence removal index is out of range"
  else
    let left, rest = split_at index vector in
    let removed, right = split_at 1 rest in
    match nth 0 removed with
    | Some value -> Ok (value, concat left right)
    | None -> assert false

let pop vector =
  match remove (length vector - 1) vector with
  | Ok (value, rest) -> Some (rest, value)
  | Error _ -> None

let slice ~start ~length:count vector =
  if start < 0 || count < 0 || start > length vector - count then
    Error "deque slice is out of range"
  else Ok (take count (drop start vector))

let to_list vector =
  let rec node_fold node acc =
    match node with
    | Leaf items -> Array.fold_right List.cons items acc
    | Branch branch -> Array.fold_right node_fold branch.children acc
  in
  match vector with Empty -> [] | Root node -> node_fold node []

(* A real structural audit rather than an estimate: it recounts every leaf, checks node widths,
   sibling heights, cached counts, and that a size table is present exactly when the layout is
   irregular (and then agrees with the children). A violation raises [Invalid_argument], which is
   unreachable through this module's own operations. *)
let statistics vector =
  match vector with
  | Empty -> { count = 0; estimated_leaves = 0; depth = 0 }
  | Root root ->
      let leaves = ref 0 in
      let fail message = invalid_arg ("RRB structural audit failed: " ^ message) in
      let rec visit ~is_root node =
        match node with
        | Leaf items ->
            let width = Array.length items in
            if width < 1 || width > branch_factor then fail "leaf size is outside 1..32";
            incr leaves;
            (width, 0)
        | Branch branch ->
            let width = Array.length branch.children in
            if width < 1 || width > branch_factor then fail "branch factor is outside 1..32";
            if is_root && width < 2 then fail "a root branch must hold at least two children";
            let total = ref 0 in
            let child_height = ref (-1) in
            Array.iter
              (fun child ->
                let child_count, height = visit ~is_root:false child in
                if !child_height = -1 then child_height := height
                else if height <> !child_height then fail "sibling heights differ";
                total := !total + child_count)
              branch.children;
            if !total <> branch.count then fail "a cached count disagrees with the children";
            if !child_height + 1 <> branch.height then fail "a cached height disagrees";
            let regular = has_regular_layout branch.children branch.height in
            (match branch.sizes with
            | None -> if not regular then fail "an irregular branch is missing its size table"
            | Some sizes ->
                if regular then fail "a regular branch carries a size table";
                if sizes <> build_sizes branch.children then fail "a size table is stale");
            (!total, branch.height)
      in
      let count, height = visit ~is_root:true root in
      { count; estimated_leaves = !leaves; depth = height + 1 }

(* --- cursors (index-based over one snapshot; edits go through the vector operations) ---------- *)

let cursor vector = { cursor_snapshot = vector; cursor_position = 0 }

let cursor_at position vector =
  if position < 0 || position > length vector then None
  else Some { cursor_snapshot = vector; cursor_position = position }

let cursor_position value = value.cursor_position
let cursor_length value = length value.cursor_snapshot
let cursor_is_at_start value = value.cursor_position = 0
let cursor_is_at_end value = value.cursor_position = cursor_length value
let cursor_peek_previous value = nth (value.cursor_position - 1) value.cursor_snapshot
let cursor_peek_next value = nth value.cursor_position value.cursor_snapshot
let cursor_move_previous value = cursor_at (value.cursor_position - 1) value.cursor_snapshot

let cursor_move_next value =
  if cursor_is_at_end value then None
  else cursor_at (value.cursor_position + 1) value.cursor_snapshot

let cursor_seek position value = cursor_at position value.cursor_snapshot

let cursor_insert element value =
  {
    cursor_snapshot = Result.get_ok (insert value.cursor_position element value.cursor_snapshot);
    cursor_position = value.cursor_position + 1;
  }

let cursor_insert_vector inserted value =
  if is_empty inserted then value
  else
    let left, right = split_at value.cursor_position value.cursor_snapshot in
    {
      cursor_snapshot = concat (concat left inserted) right;
      cursor_position = value.cursor_position + length inserted;
    }

let cursor_insert_many elements value = cursor_insert_vector (of_list elements) value

let cursor_delete_previous value =
  if cursor_is_at_start value then None
  else
    let _, snapshot = Result.get_ok (remove (value.cursor_position - 1) value.cursor_snapshot) in
    Some { cursor_snapshot = snapshot; cursor_position = value.cursor_position - 1 }

let cursor_delete_next value =
  if cursor_is_at_end value then None
  else
    let _, snapshot = Result.get_ok (remove value.cursor_position value.cursor_snapshot) in
    Some { cursor_snapshot = snapshot; cursor_position = value.cursor_position }

let cursor_replace_next element value =
  if cursor_is_at_end value then None
  else
    let snapshot = Result.get_ok (set value.cursor_position element value.cursor_snapshot) in
    Some { cursor_snapshot = snapshot; cursor_position = value.cursor_position }

let cursor_snapshot value = value.cursor_snapshot

module Builder = struct
  type 'element vector = 'element t

  (* Appends stage into a mutable tail buffer and become full leaves without touching the tree;
     indexed edits and freezes flush the staged suffix into the persistent prefix first. *)
  type 'element t = {
    mutable prefix : 'element vector;
    mutable staged_leaves : 'element node list; (* newest first *)
    mutable tail : 'element list; (* newest first *)
    mutable tail_length : int;
  }

  let create current = { prefix = current; staged_leaves = []; tail = []; tail_length = 0 }

  let staged_count builder =
    builder.tail_length + (branch_factor * List.length builder.staged_leaves)

  let length builder = length builder.prefix + staged_count builder

  let flush builder =
    if staged_count builder > 0 then begin
      let leaves =
        if builder.tail_length = 0 then builder.staged_leaves
        else Leaf (Array.of_list (List.rev builder.tail)) :: builder.staged_leaves
      in
      let staged = Root (build_level (Array.of_list (List.rev leaves))) in
      builder.prefix <- concat builder.prefix staged;
      builder.staged_leaves <- [];
      builder.tail <- [];
      builder.tail_length <- 0
    end

  let append value builder =
    builder.tail <- value :: builder.tail;
    builder.tail_length <- builder.tail_length + 1;
    if builder.tail_length = branch_factor then begin
      builder.staged_leaves <- Leaf (Array.of_list (List.rev builder.tail)) :: builder.staged_leaves;
      builder.tail <- [];
      builder.tail_length <- 0
    end

  let set index value builder =
    flush builder;
    Result.map (fun successor -> builder.prefix <- successor) (set index value builder.prefix)

  let remove index builder =
    flush builder;
    Result.map
      (fun (value, successor) ->
        builder.prefix <- successor;
        value)
      (remove index builder.prefix)

  let freeze builder =
    flush builder;
    builder.prefix
end
